### PR TITLE
Fix using schema_inference_make_columns_nullable=0

### LINKED ISSUE #67999: Improve `schema_inference_make_columns_nullable` setting
Now when `schema_inference_make_columns_nullable=0` we can still make columns `Nullable` when the data contains `Null` for this column or file metadata contains information about nullability. But sometimes there is need to avoid `Nullable` at all.

We can have 3 values for `schema_inference_make_columns_nullable`:
- `enable = 1`  - make all columns `Nullable`
- `disable = 2` - don't make any columns `Nullable`
- `auto = 0` - make columns `Nullable` if data contains `Nulls` or there is  information about nullability in metadata.

And we should backport it because the initial change was actually backward incompatible.

### PR DESCRIPTION
**Pull request by Avogar**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
After https://github.com/ClickHouse/ClickHouse/pull/61984 `schema_inference_make_columns_nullable=0` still can make columns `Nullable` in Parquet/Arrow formats. The change was backward incompatible and users noticed the changes in the behaviour. This PR makes `schema_inference_make_columns_nullable=0` to work as before (no Nullable columns will be inferred) and introduces new value `auto` for this setting that will make columns `Nullable` only if data has information about nullability.

Closes https://github.com/ClickHouse/ClickHouse/issues/67999

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by Avogar** [comment]
Let's backport it because users in the cloud also noticed the changes in the behaviour.

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit 3cfb921befa895e445e8d7b98e639015e1e41aa0 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/68298/3cfb921befa895e445e8d7b98e639015e1e41aa0/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/dd3fa7c3b697d77e9bd840b4cb6aeebd72310dea/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Core/Settings.h [hunk 1]
```diff
@@ -1116,7 +1116,7 @@ class IColumn;
     M(String, column_names_for_schema_inference, "", "The list of column names to use in schema inference for formats without column names. The format: 'column1,column2,column3,...'", 0) \
     M(String, schema_inference_hints, "", "The list of column names and types to use in schema inference for formats without column names. The format: 'column_name1 column_type1, column_name2 column_type2, ...'", 0) \
     M(SchemaInferenceMode, schema_inference_mode, "default", "Mode of schema inference. 'default' - assume that all files have the same schema and schema can be inferred from any file, 'union' - files can have different schemas and the resulting schema should be the a union of schemas of all files", 0) \
-    M(Bool, schema_inference_make_columns_nullable, true, "If set to true, all inferred types will be Nullable in schema inference for formats without information about nullability.", 0) \
+    M(UInt64Auto, schema_inference_make_columns_nullable, 1, "If set to true, all inferred types will be Nullable in schema inference. When set to false, no columns will be converted to Nullable. When set to 'auto', ClickHosue will use information about nullability from the data..", 0) \
```
Comment (alexkats) [inline src/Core/Settings.h:1119]:
```suggestion
    M(UInt64Auto, schema_inference_make_columns_nullable, 1, "If set to true, all inferred types will be Nullable in schema inference. When set to false, no columns will be converted to Nullable. When set to 'auto', ClickHouse will use information about nullability from the data.", 0) \
```

Comment (alexkats) [inline src/Core/Settings.h:1119]: Isn't `auto` a default value?

File: src/Formats/FormatSettings.h [hunk 2]
```diff
@@ -75,7 +75,7 @@ struct FormatSettings
         Raw
     };
 
-    bool schema_inference_make_columns_nullable = true;
+    UInt64 schema_inference_make_columns_nullable = true;
```
Comment (alexkats) [inline src/Formats/FormatSettings.h:78]:
```suggestion
    UInt64 schema_inference_make_columns_nullable = 1;
```

File: src/Core/Settings.h (hunk 1 above)
Comment (Avogar) [reply src/Core/Settings.h]: No, we wan't this setting to be enabled by default (as it was before this PR)

**Review by alexkats — APPROVED** [review]
LGTM, just in case, another test for schema inference has failed

**Comment by Avogar** [comment]
Need to update the tests with new better behaviour.

**Pull request merged**